geps/gep-1767/index.md

```diff
@@ -0,0 +1,701 @@
+
+## API
+This GEP proposes to add a new field `HTTPCORSFilter` to `HTTPRouteFilter`.
+If HTTPCORSFilter is set, then the gateway will generate the response of the preflight requests and send back it to the client directly.
+For the actual cross-origin request, the gateway will add CORS headers to the response before it is sent to the client.
+
+```golang
+const (
+    // HTTPRouteFilterCORS can be used to add CORS headers to an 
+    // HTTP response before it is sent to the client.
+    //
+    // Support in HTTPRouteRule: Extended
+    //
+    // Support in HTTPBackendRef: Extended
+    HTTPRouteFilterCORS HTTPRouteFilterType = "CORS"
+)
+
+type HTTPRouteFilter struct {
+    // CORS defines a schema for a filter that responds to the
+    // cross-origin request based on HTTP response header.
+    // 
+    // Support: Extended
+    //
+    // +optional
+    CORS *HTTPCORSFilter `json:"cors,omitempty"`
+}
+
+// HTTPCORSFilter defines a filter that responds to the
+// cross-origin request based on HTTP header.
+type HTTPCORSFilter struct {
+    // AllowOrigins indicates whether the response can be shared with 
+    // requested resource from the given `Origin`.
+    // 
+    // A wildcard indicates that the requests from all `Origin` are allowed.
+    //
+    // When responding to a credentialed requests, the gateway must specify 
+    // an origin in the value of the Access-Control-Allow-Origin response header, 
+    // instead of specifying the * wildcard.
+    //
+    // Input:
+    //   Origin: https://foo.example
+    //
+    // Config:
+    //   allowOrigins: ["*"]
+    //
+    // Output:
+    //   Access-Control-Allow-Origin: *
+    //
+    // For a simple cross-origin interaction, the gateway sets value of the header 
+    // Access-Control-Allow-Origin same as the `Origin` header provided by the client.
+    //
+    // The status code of a successful response to a "preflight" request is an OK status (i.e., 204 or 200). 
+    // For the "preflight" request, if the request `Origin` does not match the configured allowed origins,
+    // the gateway will return a response with error status (e.g., 403). 
+    // Moreover, the gateway will omit the relevant cross-origin response headers.
+    // Then, the client will not send actual cross-origin request.
+    // 
+    // Input:
+    //   Origin: https://foo.example
+    //
+    // Config:
+    //   allowOrigins: ["https://foo.example", "http://foo.example", "https://test.example", "http://test.example"]
+    //
+    // Output:
+    //   Access-Control-Allow-Origin: https://foo.example
+    //
+    // Support: Extended
+    // +listType=set
+    // +kubebuilder:validation:MaxItems=64
+    AllowOrigins []string `json:"allowOrigins,omitempty"`
+ 
+    // AllowCredentials indicates whether the actual cross-origin request 
+    // allows to include credentials.
+    //
+    // The only valid value for the header `Access-Control-Allow-Credentials` is true (case-sensitive).
```

Comment:
Should it really be case-sensitive? Is that what the RFC says?

Reply:
This should also include a mention that requests with `AllowCredentials` set to `true` are referred to as `credentialled requests` in other parts of the spec.

Reply:
I get it.